src/http/gateway/resolver.js

```diff
@@ -0,0 +1,114 @@
+'use strict'
+
+const mh = require('multihashes')
+const promisify = require('promisify-es6')
+const eachOfSeries = require('async/eachOfSeries')
+const debug = require('debug')
+const log = debug('jsipfs:http-gateway:resolver')
+log.error = debug('jsipfs:http-gateway:resolver:error')
+
+const html = require('./utils/html')
+const PathUtil = require('./utils/path')
+
+const INDEX_HTML_FILES = [ 'index.html', 'index.htm', 'index.shtml' ]
+
+const resolveDirectory = promisify((ipfs, path, multihash, callback) => {
+  if (!callback) {
+    callback = noop
+  }
+
+  mh.validate(mh.fromB58String(multihash))
+
+  ipfs
+   .object
+   .get(multihash, { enc: 'base58' })
+   .then((DAGNode) => {
+     const links = DAGNode.links
+     const indexFiles = links.filter((link) => INDEX_HTML_FILES.indexOf(link.name) !== -1)
+
+     // found index file in links
+     if (indexFiles.length > 0) {
+       return callback(null, indexFiles)
+     }
+
+     return callback(null, html.build(path, links))
+   })
+})
+
+const noop = function () {}
+
+const resolveMultihash = promisify((ipfs, path, callback) => {
+  if (!callback) {
+    callback = noop
+  }
+
+  const parts = PathUtil.splitPath(path)
+  const partsLength = parts.length
+
+  let currentMultihash = parts[0]
+
+  eachOfSeries(parts, (multihash, currentIndex, next) => {
+    // throws error when invalid multihash is passed
+    mh.validate(mh.fromB58String(currentMultihash))
```

Comment:
Check for CID and not multihash.

Reply:
@diasdavid Do you mean something like this 
```javascript
const CID = require('cids')
//... some code
  let currentMultihash = parts[0]
  let currentCid
  eachOfSeries(parts, (multihash, currentIndex, next) => {
    try {
      currentCid = new CID(mh.fromB58String(currentMultihash))
    } catch (e) {
      if (e) throw e
    }
```